File: 2023/9.py
with open('9input') as f:
    input = [[int(n) for n in line.split()] for line in f]


def predict_next(l):
    if all(n == 0 for n in l):
        return 0

    l2 = [x - y for x, y in zip(l[1:], l[:-1])]
    return l[-1] + predict_next(l2)

print('Part 1:', sum(predict_next(l) for l in input))


def predict_previous(l):
    if all(n == 0 for n in l):
        return 0

    l2 = [x - y for x, y in zip(l[1:], l[:-1])]
    return l[0] - predict_previous(l2)

print('Part 1:', sum(predict_previous(l) for l in input))

File: 2025/9.clj
(require '[clojure.string :as str])

(defn parse-input [input-str]
  (for [l (str/split-lines input-str)]
    (mapv parse-long (str/split l #","))))

(defn all-squares [coords]
  (loop [coords coords
         squares []]
    (if (not (seq (rest coords)))
      squares
      (recur (rest coords)
             (apply conj squares
                    (for [c2 (rest coords)]
                      [(first coords) c2]))))))

(defn square-area [[[x1 y1] [x2 y2]]]
  (* (abs (+ 1 (- x1 x2)))
     (abs (+ 1 (- y1 y2)))))

(println "Part 1:"
         (->> (parse-input (slurp "9input"))
              all-squares
              (map square-area)
              (reduce max)))
